_posts/2018-06-12-making-cloudformation-awesome.md

```diff
@@ -0,0 +1,238 @@
+acme-orders-api
+    acme-orders-service.yml (ECS definitions, target group, ALB routing rules, Imports necessary stuff)
+acme-customers-api
+    acme-customers-service.yml (ECS definitions, target group, ALB routing rules, Imports necessary stuff)    
+```
+
+Using shared ALB like this requires some additional tweaks, i.e sharing routing priority numbers between services, but that is an implementation detail. Rule of the thumb is that you should have just the right amount of stacks and dependency graph must be unidirectional. So kids! Remember not to use `!Import` in the infra stack! 😘
+
+It's super important to understand in which stack you should add your new resource in a multi-stack scenario, because you can really easily make circular dependencies between stacks. Think for instance if you have created separate stacks for [CloudFront](https://aws.amazon.com/cloudfront/) and [S3 buckets](https://aws.amazon.com/s3/). Your CloudFront stack needs to know S3 bucket domains and your S3 bucket policy is S3 stack needs to know CloudFront [Origin Access Identity](https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/private-content-restricting-access-to-s3.html). There you have a circular dependency and you probably spend loadsa time debugging and possibly even suffer some downtime when stuff hits the fan.
+
+But the good news is that when you just bake both CloudFront and S3 stuff into the same stack - it just works.
+
+Bottom line is that, when having multiple stacks, you should be really careful when selecting a stack for your new resource. It's a good idea to have well maintained visualisation of the stacks and exports/imports. Don't be afraid to make a loose coupling between resources when needed! It's totally ok to just copy resource references directly as a parameters to other stacks instead of exporting and importing them. Exporting is a really powerful tool and it makes really hard coupling between the stacks. Try to minimise the amount of stacks. Less is more.
+
+What about the different environments then?
+
+### 3. Parametrise Right Things Instead of All Things
+
+CloudFormation supports [parameters](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/parameters-section-structure.html) for stacks which can be given as a JSON file, from [SSM Parameter Store](https://docs.aws.amazon.com/systems-manager/latest/userguide/systems-manager-paramstore.html) or directly with AWS CLI. This is nice, because I would hate to maintain `acme-app-dev.yml` and `acme-app-prod.yml` - both thousands of lines long.
+
+My CloudFormation parametrisation principle is really easy to follow:
+
+1. Parametrise attributes that actually are different in your environments
+2. Don't parametrise things that you think can eventually change in your environments
+3. When thing changes which is not parametrised, see if it is different per environment. If it is different, add it as a parameter, and if not, leave it to the template.
```

Comment:
When a thing changes that is not parameterised, ...  If it is (different-sanan toiston voi jättää), add ...
Uus lause ehkä loppuun:  And if not, ...